### PR TITLE
Improve the parallel part of the K-point module

### PR DESCRIPTION
**Pull request by jwsyzy**
### What's changed?
* remove unnecessary calculation in `ibz_kpoint` in `klist.cpp`
* improve `gathervec` and `pool_collection` in `parallel_kpoints.cpp`
* use `vector` to displace `new` and `delete`
* remove the direct use of `GlobalV` in class `Parallel_Kpoints`
* remove unused variable `nproc_pool`
* add definition to `startpro_pool` and move it to private

### CODE REVIEW

File: source/module_cell/parallel_kpoints.h [hunk 1]
```diff
@@ -37,24 +37,35 @@ class Parallel_Kpoints
 #endif
 
     // information about pool, dim: GlobalV::KPAR
-    int* nproc_pool = nullptr;
-    int* startpro_pool = nullptr;
+    //int* nproc_pool = nullptr;    it is not used
+    //int* startpro_pool = nullptr;
+    std::vector<int> startpro_pool;
```
Comment (dyzheng) [inline source/module_cell/parallel_kpoints.h]: The new variables in Class should not be public members, please refactor to use `set_xx()` and `get_xx()` interfaces to access them?

File: source/module_cell/parallel_kpoints.h [hunk 2]
```diff
@@ -37,24 +37,35 @@ class Parallel_Kpoints
...
     int kpar = 0;               // number of pools
 
     // information about which pool each k-point belongs to,
-    int* whichpool = nullptr; // whichpool[k] : the pool which k belongs to, dim: nkstot_np
+    std::vector<int> whichpool; // whichpool[k] : the pool which k belongs to, dim: nkstot_np
+    
     int nkstot_np = 0;        // number of k-points without spin, kv.set_nkstot(nkstot_np) * nspin(1 or 2)
     int nks_np = 0;           // number of k-points without spin in the present pool
 
+    int my_pool = 0;          // the pool index of the present processor
```
Comment (dyzheng) [inline source/module_cell/parallel_kpoints.h:60]: These four variables should be private without access interface.

File: source/module_cell/klist.cpp [hunk 3]
```diff
@@ -668,14 +668,16 @@ void K_Vectors::ibz_kpoint(
         std::string bkbrav_name;
         ModuleBase::Vector3<double>gk01 = gk1, gk02 = gk2, gk03 = gk3;
 
-        ModuleBase::Matrix3 b_optlat = symm.optlat.Inverse().Transpose();
-        //search optlat after using reciprocity relation
-        ModuleBase::Vector3<double> gb01(b_optlat.e11, b_optlat.e12, b_optlat.e13);
-        ModuleBase::Vector3<double> gb02(b_optlat.e21, b_optlat.e22, b_optlat.e23);
-        ModuleBase::Vector3<double> gb03(b_optlat.e31, b_optlat.e32, b_optlat.e33);
+        // it's not necessary to calculate gb01, gb02, gb03,
```
Comment (dyzheng) [inline source/module_cell/klist.cpp:671]: You can be safely to delete useless code

File: source/module_cell/parallel_kpoints.h [hunk 4]
```diff
@@ -37,24 +37,38 @@ class Parallel_Kpoints
...
+    std::vector<int> startk_pool; // the first k-point in each pool
 
     // information about which pool each k-point belongs to,
-    int* whichpool = nullptr; // whichpool[k] : the pool which k belongs to, dim: nkstot_np
+    std::vector<int> whichpool; // whichpool[k] : the pool which k belongs to, dim: nkstot_np
+    
     int nkstot_np = 0;        // number of k-points without spin, kv.set_nkstot(nkstot_np) * nspin(1 or 2)
     int nks_np = 0;           // number of k-points without spin in the present pool
+    
+    int get_startpro_pool(const int& pool) const
```
Comment (dyzheng) [inline source/module_cell/parallel_kpoints.h]: Public interface in class should have annotation , which should be doxygen format(/// or /**...*/)

Comment (dyzheng) [reply source/module_cell/parallel_kpoints.h]: New code should follow this criteria, existed code can be skipped

**Review by mohanchen — APPROVED** [review]

**Review by mohanchen — APPROVED** [review]

**Pull request merged**